mapper/pojo-base/src/main/java/org/hibernate/search/mapper/pojo/massindexing/loader/MassIndexingTypeLoadingContext.java

```diff
@@ -0,0 +1,84 @@
+/*
+ * Hibernate Search, full-text search for your domain model
+ *
+ * License: GNU Lesser General Public License (LGPL), version 2.1 or later
+ * See the lgpl.txt file in the root directory or <http://www.gnu.org/licenses/lgpl-2.1.html>.
+ */
+package org.hibernate.search.mapper.pojo.massindexing.loader;
+
+import java.util.Map;
+import org.hibernate.search.mapper.pojo.massindexing.loader.dsl.MassIndexingTypeBatchLoader;
+import org.hibernate.search.mapper.pojo.massindexing.loader.dsl.MassIndexingTypeBatchingStep;
+
+/**
+ * The DSL entry point passed to consumers in
+ *
+ * @param <E> The resulting entity type (output)
+ * @param <C> The contextural indexer information
+ */
+public interface MassIndexingTypeLoadingContext<E, C extends MassIndexerContext> {
+
+	/**
+	 * Get the contextural indexer information.
+	 *
+	 * @return contextural indexer information.
+	 */
+	C indexer();
+
+	/**
+	 * @return the identifier of the tenant.
+	 */
+	String tenantId();
+
+	/**
+	 * Get the context data for this indexing query.
+	 *
+	 * @return context data.
+	 */
+	Map<Object, Object> contextData();
+
+	/**
+	 * Set the option for inxexing process.
+	 *
+	 * @param value The param value.
+	 * @return {@code this} for method chaining.
+	 */
+	MassIndexingTypeLoadingContext<E, C> totalCount(long value);
+
+	/**
+	 * Get the in progress indexing process,
+	 *
+	 * @return true if process is active.
+	 */
+	boolean active();
```

Comment:
It is hard to assume that it is always about a transaction. This is the signal to break the interception loop. A programmer building an interception will be able to signal an interrupt, no matter what. It could be a transaction. 

```java
public final class TransactionInterceptors {
	private static final Log log = LoggerFactory.make( Log.class, MethodHandles.lookup() );

	private TransactionInterceptors() {
	}

	public static LoadingInterceptor<? extends TransactionContext> withJTATransaction(TransactionManager tm) {
		return new JavaBeanTransactionInterceptor( new JavaBeanLoadingTransactionalContext( tm ), false );
	}

	public static LoadingInterceptor<? extends TransactionContext> withJTAUncommittedTransaction(TransactionManager tm) {
		return new JavaBeanTransactionInterceptor( new JavaBeanLoadingTransactionalContext( tm ), true );
	}

	private static class JavaBeanTransactionInterceptor implements LoadingInterceptor<TransactionContext> {

		private final JavaBeanLoadingTransactionalContext transactionalContext;
		private final boolean rolbackAlweys;

		public JavaBeanTransactionInterceptor(JavaBeanLoadingTransactionalContext transactionalContext, boolean rolbackAlweys) {
			this.transactionalContext = transactionalContext;
			this.rolbackAlweys = rolbackAlweys;
		}

		@Override
		public void intercept(LoadingInvocationContext<TransactionContext> ictx) throws Exception {
			TransactionContext loadingContext = ictx.loadingContext();
			boolean wrapInTransaction = transactionalContext.wrapInTransaction();
			Integer transactionTimeout = loadingContext.transactionTimeout();

			if ( wrapInTransaction ) {
				try {
					if ( transactionTimeout != null ) {
						transactionalContext.transactionManager().setTransactionTimeout( transactionTimeout );
					}
					ictx.active( transactionalContext::transactionInProgress );
					ictx.proceed( next -> {
						transactionalContext.transactionManager().begin();
						next.proceed();
						if ( rolbackAlweys ) {
							transactionalContext.transactionManager().rollback();

						}
						else {
							transactionalContext.transactionManager().commit();
						}
					} );
				}
				// Just let runtime exceptions fall through
				catch (NotSupportedException | SystemException | RollbackException | HeuristicMixedException
						| HeuristicRollbackException e) {
					throw log.massIndexingTransactionHandlingException( e.getMessage(), e );
				}
				finally {
					rollback();
				}
			}
			else {
				ictx.proceed();
			}
		}

		private void rollback() {
			boolean wrapInTransaction = transactionalContext.wrapInTransaction();
			if ( wrapInTransaction ) {
				try {
					transactionalContext.transactionManager().rollback();
				}
				catch (SystemException e) {
					log.errorRollingBackTransaction( e.getMessage(), e );
				}
			}
		}
	}
}
```

Reply:
The transaction interceptor "TransactionInterceptors.JavaBeanTransactionInterceptor" in JavaBean needs this to control the loop break. The idea is to isolate user-added interceptions from entity loading. 

```java
			if ( wrapInTransaction ) {
				try {
					if ( transactionTimeout != null ) {
						transactionalContext.transactionManager().setTransactionTimeout( transactionTimeout );
					}
					ictx.active( transactionalContext::transactionInProgress );
					ictx.proceed( next -> {
						transactionalContext.transactionManager().begin();
						next.proceed();
						if ( rolbackAlweys ) {
							transactionalContext.transactionManager().rollback();

						}
						else {
							transactionalContext.transactionManager().commit();
						}
					} );
				}
				// Just let runtime exceptions fall through
				catch (NotSupportedException | SystemException | RollbackException | HeuristicMixedException
						| HeuristicRollbackException e) {
					throw log.massIndexingTransactionHandlingException( e.getMessage(), e );
				}
				finally {
					rollback();
				}
			}
```

Reply:
Moved to "MassIndexingThreadContext".

Reply:
In mapper JavaBean as in mapper ORM required to break the loop. In java bean used during JPA transaction interception. 

```java
public final class TransactionInterceptors {
..................
	public static LoadingInterceptor<? extends TransactionOptions> withJTATransaction(TransactionManager tm) {
		return new JavaBeanTransactionInterceptor( new JavaBeanLoadingTransactionalContext( tm ), false );
	}

	public static LoadingInterceptor<? extends TransactionOptions> withJTAUncommittedTransaction(TransactionManager tm) {
		return new JavaBeanTransactionInterceptor( new JavaBeanLoadingTransactionalContext( tm ), true );
	}

	private static class JavaBeanTransactionInterceptor implements LoadingInterceptor<TransactionOptions> {
...................
		@Override
		public void intercept(LoadingInvocationContext<TransactionOptions> ictx) throws Exception {
			TransactionOptions options = ictx.options();
			boolean wrapInTransaction = transactionalContext.wrapInTransaction();
			Integer transactionTimeout = options.transactionTimeout();

			if ( wrapInTransaction ) {
				try {
					if ( transactionTimeout != null ) {
						transactionalContext.transactionManager().setTransactionTimeout( transactionTimeout );
					}
					ictx.active( transactionalContext::transactionInProgress );
					ictx.proceed( next -> {
						transactionalContext.transactionManager().begin();
						next.proceed();
						if ( rolbackAlweys ) {
							transactionalContext.transactionManager().rollback();

						}
						else {
							transactionalContext.transactionManager().commit();
						}
					} );
				}
				// Just let runtime exceptions fall through
				catch (NotSupportedException | SystemException | RollbackException | HeuristicMixedException
						| HeuristicRollbackException e) {
					throw log.massIndexingTransactionHandlingException( e.getMessage(), e );
				}
				finally {
					rollback();
				}
			}
			else {
				ictx.proceed();
			}
```